### PR TITLE
AO3-5307 update series bookmarks when work is added to series

### PR DESCRIPTION
**Pull request by WendyBeth**
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added tests for any changed functionality?
* [x] Have you added the JIRA issue number as the *first* thing in your pull request title (eg: `AO3-1234 Fix thing`)
* [x] Have you updated the JIRA issue with the information below?

## Issue

https://otwarchive.atlassian.net/browse/AO3-5307

## Purpose

What does this PR do?

This PR fixes a bug where, if a series has works all with fandom A and is bookmarked, and then a work is added to that series with fandom B, then the series bookmark should show up for bookmarks tagged with Fandom B. It also ensures that the updated_at order is preserved.

## Testing

How can the Archive's QA team verify that this is working as you intended? (If you have access, please copy this into the JIRA ticket for them!)

Create a series with a work with any Fandom A, the bookmark the work. Go the Fandom A bookmarks to ensure the series bookmark is listed there. Add a work with Fandom B to the series, and then go to the Fandom B bookmarks and ensure that the series bookmark is listed there as well. Add other bookmarks, and then go through the same process again and ensure that bookmarks for series which have just had new works appended to them show up first when sorting the bookmarks by 'Updated Date'.

## References

Are there any other relevant issues / PRs / mailing lists discussions related to this?

## Credit

*What name do you want us to use to credit your work in the [Archive of Our Own's Release Notes](http://archiveofourown.org/admin_posts?tag=1)?*

Wendy Randquist - Littlelines

*What pronouns do you prefer (she/her, he/him, zie/hir etc)?*

she/her

(if you do not fill this in, we will use your GitHub account name and will use "they" to avoid making assumptions about your gender)

### CODE REVIEW

**Comment by sarken** [comment]
This looks good to me! Just waiting for someone who knows the old indexing code better than I do to confirm before we merge.

**Comment by WendyBeth** [comment]
I apologize for my delayed response. I've changed the two lines @tickinginstant pointed out. In retrospect I think I was approaching the "bookmark all the works' indexes too" idea with a sense of caution - I didn't want to accidentally miss anything. But it's probably unnecessary.

File: app/models/serial_work.rb [hunk 1]
```diff
@@ -21,4 +23,15 @@ def delete_empty_series
       self.series.destroy
     end
   end
+
+  # Ensure series bookmarks are reindexed when a new work is added to a series
+  def update_series_index
+    if $rollout.active?(:start_new_indexing)
+      Series.all.each(&:enqueue_to_index)
```
Comment (tickinginstant) [inline app/models/serial_work.rb]: Sorry, I think I gave you the wrong idea here with my phrasing -- it's not necessary to reindex all *series*, but it is necessary to reindex all *bookmarks* for the series that was modified, just as in the old indexing code below.

File: app/models/serial_work.rb [hunk 2]
```diff
@@ -21,4 +23,16 @@ def delete_empty_series
...
+
+  # Ensure series bookmarks are reindexed when a new work is added to a series
+  def update_series_index
+    if $rollout.active?(:start_new_indexing)
+      series.enqueue_to_index
+      series.bookmarks.each(&:enqueue_to_index)
+    end
+
+    unless $rollout.active?(:stop_old_indexing)
+      IndexQueue.enqueue_ids(Bookmark, series.bookmarks.pluck(:id), :background)
```
Comment (redsummernight) [inline app/models/serial_work.rb]: Old indexing code looks good to me. It's similar to:

https://github.com/otwcode/otwarchive/blob/f274643b0dcca0f4a21f00d23c0f59b2b8096ffd/app/models/redis_search_index_queue.rb#L42

which predates the upgrade.

**Pull request merged**